lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,313 @@
+    , isShared
+    , keyHashFromAccXPubIx
+    , constructAddressFromIx
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( Cosigner (..)
+    , KeyHash (..)
+    , Script (..)
+    , ScriptTemplate (..)
+    , ValidationLevel (..)
+    , foldScript
+    , toScriptHash
+    , validateScriptTemplate
+    )
+import Cardano.Address.Style.Shelley
+    ( Credential (..)
+    , delegationAddress
+    , deriveMultisigPublicKey
+    , hashKey
+    , liftXPub
+    , mkNetworkDiscriminant
+    )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationPrefix (..)
+    , DerivationType (..)
+    , Index (..)
+    , NetworkDiscriminant (..)
+    , SoftDerivation
+    , WalletKey (..)
+    , deriveVerificationKey
+    , hashVerificationKey
+    )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( coinTypeAda )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( AddressPoolGap )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..), AddressState (..) )
+import Control.DeepSeq
+    ( NFData )
+import Data.Either
+    ( isRight )
+import Data.Map.Strict
+    ( Map )
+import Data.Maybe
+    ( fromMaybe )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Address as CA
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+                                 State
+-------------------------------------------------------------------------------}
+
+-- | A state to keep track of script templates, account public keys of other cosigners,
+-- | verification keys used and unused fitting in the address pool gap,
+-- | and script addresses unused/used with the corresponding indices.
+data SharedState (n :: NetworkDiscriminant) k = SharedState
+    { shareStateAccountKey :: k 'AccountK XPub
+        -- ^ Reward account public key associated with this wallet
+    , shareStateDerivationPrefix :: !DerivationPrefix
+        -- ^ Derivation path prefix from a root key up to the account key
+    , shareStateGap :: !AddressPoolGap
+        -- ^ Number of first keys that are used to produce candidate addresses
+    , shareStatePaymentTemplate :: !ScriptTemplate
+        -- ^ Script template together with a map of account keys and cosigners
+        -- for payment credential
+    , shareStateDelegationTemplate :: !(Maybe ScriptTemplate)
+        -- ^ Script template together with a map of account keys and cosigners
+        -- for staking credential. If not specified then the same template as for
+        -- payment is used
+    , shareStateOurAddresses :: !(Map Address (Index 'Soft 'ScriptK, AddressState))
+        -- ^ Our addresses meaning the addresses containing our verification key hashes
+        -- (if discovered) and represented here by corresponding indices of verification
+        -- keys present in the given address.
```

Comment:
We'll likely need something more like we have in `AddressPool`:

```
        :: !(Map
                (KeyFingerprint "payment" key)
                (Index 'Soft 'AddressK, AddressState)
            )
```

because, an address belongs to the wallet only if the payment part is ours. The delegation part may be totally different. 

Shall we consider reviewing a bit `AddressPool` to allow for re-use here :thinking: ? The code exists and is battle-tested, so it may be good to not redo all that non-trivial logic.

Reply:
I'm curious where you will go, because for #2521 I wonder if it couldn't be generalised to:

```haskell
:: !(Map
    (KeyFingerprint (s :: Symbol) key)
    (Index 'Soft 'AddressK, keyState)
)
```
to also work with stake keys

Reply:
yes, thanks, I am aware that `AddressPool` should be somehow reused here, the first step is here https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/a69d4db860f8f949f588b0533b06260c06a8966a But maybe it will be possible to merry this with what is in Sequential.